packages/ckeditor5-engine/src/view/observer/inputobserver.ts

```diff
@@ -68,32 +70,50 @@ export default class InputObserver extends DomEventObserver<'beforeinput'> {
 		// in the fake selection container.
 		if ( viewDocument.selection.isFake ) {
 			// Future-proof: in case of multi-range fake selections being possible.
-			targetRanges = [ ...viewDocument.selection.getRanges() ];
+			targetRanges = Array.from( viewDocument.selection.getRanges() );
 
 			// @if CK_DEBUG_TYPING // if ( window.logCKETyping ) {
 			// @if CK_DEBUG_TYPING // 	console.info( '%c[InputObserver]%c using fake selection:',
-			// @if CK_DEBUG_TYPING // 		'color: green;font-weight: bold', 'font-weight:bold', targetRanges
+			// @if CK_DEBUG_TYPING // 		'color: green;font-weight: bold', 'font-weight:bold', targetRanges,
+			// @if CK_DEBUG_TYPING // 		viewDocument.selection.isFake ? 'fake view selection' : 'fake DOM parent'
 			// @if CK_DEBUG_TYPING // 	);
 			// @if CK_DEBUG_TYPING // }
-		} else {
+		} else if ( domTargetRanges.length ) {
 			targetRanges = domTargetRanges.map( domRange => {
-				return view.domConverter.domRangeToView( domRange );
+				return view.domConverter.domRangeToView( domRange )!;
 			} );
 
 			// @if CK_DEBUG_TYPING // if ( window.logCKETyping ) {
 			// @if CK_DEBUG_TYPING // 	console.info( '%c[InputObserver]%c using target ranges:',
 			// @if CK_DEBUG_TYPING // 		'color: green;font-weight: bold', 'font-weight:bold', targetRanges
 			// @if CK_DEBUG_TYPING // 	);
 			// @if CK_DEBUG_TYPING // }
+		} else if ( env.isAndroid ) {
+			const domSelection = ( domEvent.target as HTMLElement ).ownerDocument.defaultView!.getSelection()!;
+
+			targetRanges = Array.from( view.domConverter.domSelectionToView( domSelection ).getRanges() );
+
+			// @if CK_DEBUG_TYPING // if ( window.logCKETyping ) {
+			// @if CK_DEBUG_TYPING // 	console.info( '%c[InputObserver]%c using selection ranges:',
+			// @if CK_DEBUG_TYPING // 		'color: green;font-weight: bold', 'font-weight:bold', targetRanges
+			// @if CK_DEBUG_TYPING // 	);
+			// @if CK_DEBUG_TYPING // }
 		}
 
-		this.fire( domEvent.type, domEvent, {
-			data,
-			dataTransfer,
-			isComposing: domEvent.isComposing,
-			targetRanges,
-			inputType: domEvent.inputType
-		} );
+		if ( env.isAndroid && domEvent.inputType == 'insertCompositionText' && data && data.endsWith( '\n' ) ) {
```

Comment:
Missing a comment explaining what is this case. Especially that we even had a ticket for that so it's easy to explain why it's needed.